File: 115-distinct-subsequences/distinct-subsequences.java
class Solution {
    Integer[][] dp;
    public int numDistinct(String s, String t) {
        dp=new Integer[s.length()][t.length()];
        return coin(s,t,0,0);
    }
    public int coin(String s,String t,int i,int j){
        if(j==t.length()){
            return 1;
        }
        if(i==s.length()){
            return 0;
        }
        if(dp[i][j]!=null){
            return dp[i][j];
        }
        int inc=0;
        if(s.charAt(i)==t.charAt(j)){
            inc=coin(s,t,i+1,j+1);
        }
        int exc=coin(s,t,i+1,j);
        return dp[i][j]=inc+exc;
    }
}

File: 115-distinct-subsequences/distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int n=s.length();
        int m=t.length();
        vector<vector<int>> dp(n,vector<int>(m,-1));
        return count(s,t,0,0,dp);
    }
    int count(string &s,string &t,int i,int j,vector<vector<int>> &dp){
        if(j==t.length()){
            return 1;
        }
        if(i==s.length()){
            return 0;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        int ans=0;
        if(s[i]==t[j]){
            ans=count(s,t,i+1,j+1,dp);
        }
        ans+=count(s,t,i+1,j,dp);
        return dp[i][j]=ans;
    }
};